include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+}
+
+irr_glsl_AnisotropicMicrofacetCache getSmoothMicrofacetCache(in float NdotV, in float NdotL)
+{
+	irr_glsl_AnisotropicMicrofacetCache uf;
+	const float d = 1e-4;//some delta to avoid zeroes... not a perfect solution but for now..
+	uf.isotropic.VdotH = NdotV-d;
+	uf.isotropic.LdotH = NdotL-d;
+	uf.isotropic.NdotH = d;
+	uf.isotropic.NdotH2 = d*d;
+	uf.TdotH = 1.0-d;
+	uf.BdotH = 1.0-d;
+
+	return uf;
+}
+
+irr_glsl_LightSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand, out vec3 out_remainder, out float out_pdf, out instr_t out_generatorInstr, out irr_glsl_AnisotropicMicrofacetCache out_uf)
+{
+	uint ix = 0u;
+	instr_t instr = irr_glsl_MC_fetchInstr(stream.offset);
+	uint op = instr_getOpcode(instr);
+	float weight = 1.0;
+	vec3 u = rand;
+
+	instr_execute_SET_GEOM_NORMAL();
+	while (!op_isBXDF(op))
+	{
+#ifdef OP_BLEND
+		if (op==OP_BLEND) {
+			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+			params_t params = instr_getParameters(instr, bsdf_data);
+			float w = params_getBlendWeight(params);
+			float rcpChoiceProb;
+			bool choseRight = irr_glsl_partitionRandVariable(w, u.z, rcpChoiceProb);
+
+			uint right = instr_getRightJump(instr);
+			ix = choseRight ? right:(ix+1u);
+			weight *= choseRight ? (1.0-w):w;
+		}
+		else 
+#endif //OP_BLEND
+		{
+#ifdef OP_SET_GEOM_NORMAL
+			if (op==OP_SET_GEOM_NORMAL)
+			{
+				instr_execute_SET_GEOM_NORMAL();
+			} else 
+#endif //OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			if (op==OP_BUMPMAP)
+			{
+				mat2x4 srcs = instr_fetchSrcRegs(instr);
+				instr_execute_BUMPMAP(instr, srcs);
+			} else
+#endif //OP_BUMPMAP
+			{}
+			ix = ix + 1u;
+		}
+
+		instr = irr_glsl_MC_fetchInstr(stream.offset+ix);
+		op = instr_getOpcode(instr);
+	}
+
+	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	params_t params = instr_getParameters(instr, bsdf_data);
+	//speculatively
+	const float ax = params_getAlpha(params);
+	const float ax2 = ax*ax;
+	const float ay = params_getAlphaV(params);
+	const float ay2 = ay*ay;
+	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data,op);
+	const bool is_bsdf = !op_isBRDF(op);
+	const vec3 refl = params_getReflectance(params);
+
+#ifndef NO_TWOSIDED
+	bool ts_flag = false;
+	handleTwosided_interactionOnly(ts_flag, instr);
+#endif //NO_TWOSIDED
+
+	const bool is_plastic = op==OP_PLASTIC;
+
+#ifdef OP_DIFFUSE
+
+#define OP_DIFFUSE_ALIAS OP_DIFFUSE
+
+#else
+
+#define OP_DIFFUSE_ALIAS 0xdeadbeefu
+
+#endif
+
+	float pdf;
+	vec3 rem;
+	uint ndf = instr_getNDF(instr);
+	irr_glsl_LightSample s;
+	irr_glsl_AnisotropicMicrofacetCache uf;
+
+	// the if/else statements below cannot be done as switch/case because it checks both `op` and `ndf`
+#ifdef OP_PLASTIC
+	float plastic_spec_weight;
+	if (is_plastic) {
+		vec3 fresnel = irr_glsl_fresnel_dielectric(ior[0],currInteraction.isotropic.NdotV);
+		float ws = dot(fresnel, CIE_XYZ_Luma_Y_coeffs);
+		bool choseDiffuse = u.z>=ws;
+		op = choseDiffuse ? OP_DIFFUSE_ALIAS : op;
+		plastic_spec_weight = ws;
```

Comment:
use `irr_glsl_partitionRandVariable` instead

you need to just divide the `out_pdf` by the `rcpChoiceProb` and multiply the future `out_rem` by `rcpChoiceProb`

**this is the only exception to not multiplying the `out_rem` by the pdf of generator choice (because the probability here will not match a `vec3` fresnel exactly)**

Reply:
Also you have a bug because `u.z` does not get paritioned/rescaled